cwl/src/main/scala/cwl/CommandLineTool.scala

```diff
@@ -40,23 +43,58 @@ case class CommandLineTool private(
+    for {
+      callableTaskDefinition <- taskDefinition(parentWorkflow = None, validator).toEither
+      executableTaskDefinition = ExecutableTaskDefinition.tryApply(callableTaskDefinition).toEither
+      executable <- CwlExecutableValidation.buildWomExecutable(executableTaskDefinition, inputFile)
+    } yield executable
+  }
+
+  private def validateRequirementsAndHints(parentWorkflow: Option[Workflow], validator: Requirement => ErrorOr[Requirement]): ErrorOr[List[Requirement]] = {
+    import cats.instances.list._
+    import cats.syntax.traverse._
+
+    val allRequirements = requirements.toList.flatten ++ parentWorkflow.toList.flatMap(_.allRequirements)
+    // All requirements must validate or this fails.
+    val errorOrValidatedRequirements: ErrorOr[List[Requirement]] = allRequirements traverse validator
+
+    errorOrValidatedRequirements map { validRequirements =>
+      // Only Requirement hints, everything else is thrown out.
+      // TODO CWL don't throw them out but pass them back to the caller to do with as the caller pleases.
+      val hintRequirements = hints.toList.flatten.flatMap { _.select[Requirement] }
+      val parentHintRequirements = parentWorkflow.toList.flatMap(_.allHints)
+
+      // Throw out invalid Requirement hints.
+      // TODO CWL pass invalid hints back to the caller to do with as the caller pleases.
+      val validHints = (hintRequirements ++ parentHintRequirements).collect { case req if validator(req).isValid => req }
+      validRequirements ++ validHints
+    }
+  }
 
-  def taskDefinition: CallableTaskDefinition = {
+  def taskDefinition(parentWorkflow: Option[Workflow], validator: Requirement => ErrorOr[Requirement]): ErrorOr[CallableTaskDefinition] = {
 
-    val id = this.id
+    validateRequirementsAndHints(parentWorkflow, validator) map { requirementsAndHints =>
+      val id = this.id
 
-    val commandTemplate: Seq[CommandPart] = baseCommand.toSeq.flatMap(_.fold(BaseCommandToCommandParts)) ++
-      arguments.toSeq.flatMap(_.map(_.fold(ArgumentToCommandPart))) ++
-      CommandLineTool.orderedForCommandLine(inputs).map(InputParameterCommandPart.apply)
+      val commandTemplate: Seq[CommandPart] = baseCommand.toSeq.flatMap(_.fold(BaseCommandToCommandParts)) ++
+        arguments.toSeq.flatMap(_.map(_.fold(ArgumentToCommandPart))) ++
+        CommandLineTool.orderedForCommandLine(inputs).map(InputParameterCommandPart.apply)
 
-    val runtimeAttributes: RuntimeAttributes = RuntimeAttributes(Map.empty[String, WomExpression])
+      val dockerRequirement = requirementsAndHints.toStream flatMap { _.select[DockerRequirement] } headOption
+      val attributeMap = for {
+        requirement <- dockerRequirement
+        // TODO CWL this only supports `dockerPull`. Not clear from the spec what to do if multiple attributes are specified.
+        pull <- requirement.dockerPull
+      } yield "docker" -> ValueAsAnExpression(WomString(pull))
```

Comment:
The other occurrences of the `"docker"` literal used as a runtime attribute key are in packages that aren't visible to `cwl`. Not sure where the right place for these would be, maybe `common`?

Reply:
IMO I feel like it could even be in `core`. I see `common` more as a bunch of mostly non business logic, scala related code.

Reply:
Oh right, yes. I meant `core` not `common` née `lenthall`.